neo/IO/Caching/FIFOSet.cs

```diff
@@ -1,16 +1,30 @@
 using System;
 using System.Collections;
 using System.Collections.Generic;
-using System.Collections.Specialized;
-using System.Linq;
 
 namespace Neo.IO.Caching
 {
     internal class FIFOSet<T> : IEnumerable<T> where T : IEquatable<T>
     {
+        class Entry
```

Comment:
Why change this class?

Reply:
Is faster than the `OrderedDictionary`

Reply:
Is this a linked list?

Reply:
Yes

Reply:
Then the performance of `Contains()` is low. This is very important. Otherwise we can use `System.Collections.Generic.LinkedList<T>` directly.

Reply:
> In my benchmarks `Contains` was faster than the previous version. But we can try more benchmarks with LinkedList too

After testing it shows that func contain is not performing very well. TPS is not improved.

Reply:
Let's revert this change.

Reply:
> After testing it shows that func Contains is not performing very well. This func takes almost 15 seconds in a block period.

Which version? with linkedList I obtained a better result comparing with the original version

Reply:
> > After testing it shows that func Contains is not performing very well. This func takes almost 15 seconds in a block period.
> 
> Which version? with linkedList I obtained a better result comparing with the original version

This is the line that perfroms not good:

`UInt256[] hashes = payload.Hashes.Where(p => knownHashes.Add(p) && !sentHashes.Contains(p)).ToArray();`

And this is the definition of func Contains:

`public bool Contains(T item)
        {
            var current = _firstEntry;

            while (current != null)
            {
                if (current.Value.Equals(item)) return true;
                current = current.Next;
            }

            return false;
        }`

It's defined in FIFOSet.cs.

Reply:
According to my tests, this `Contains` is faster than the previous version.

Reply:
> According to my tests, this `Contains` is faster than the previous version.

But the computation complexity of this function is O(n), could we use data structure like dictionary or hashmap to change its complexity to O(1)?

Reply:
> According to my tests, this `Contains` is faster than the previous version.

Have you tested occasions when the FIFOSet be very large? In this occasion, when knownHashes trys to add payload hashes into itself, it will firstly check whether the specified hash already exists. So if knowhashes is very large and transactions keep pouring in, time consumed by this line can be rather long.

Reply:
This check was done in the previous version also

Reply:
https://github.com/neo-project/neo/blob/cd85039a1e9a671f1a5ca0f93087b6de2606febe/neo/IO/Caching/FIFOSet.cs#L27

Reply:
> This check was done in the previous version also

But I think in previous version the computation complexity of func contains is O(1)

Reply:
Agree with @Qiao-Jin. I can't imagine that linked lists will be faster than hash tables.